### ISSUE TITLE
HMR

### ISSUE THREAD

**Issue by konsumer** [vadimdemedes/ink#159]
I am interested in HMR (hot module reloading) for ink, so it can keep state and reload components on file-change. I'm not sure where to start. It seems like [react-hot-loader](https://github.com/gaearon/react-hot-loader) is pretty focused on webpack & web-oriented workflows. Any suggestions would be welcomed.

Might be sort of related to #15

**Comment by konsumer** [comment]
I started with a kind of naive approach, just to see if I could rerender on change.

I have an `App` that I want to fill the screen, and get re-rendered if a file changes (and eventually when terminal size changes.) I am using hooks and an HOC, because this illustrates how I will really use it.

Here is my `App`:

```js
import React, { useState, useEffect }  from 'react'
import { Box, StdoutContext } from 'ink'
import Gradient from 'ink-gradient'
import BigText from 'ink-big-text'

// HOC to give your component stdout
export const withStdout = (Component) => props => <StdoutContext.Consumer>{({ stdout }) => <Component stdout={stdout} {...props} />}</StdoutContext.Consumer>

// simple demo-counter hook
export const useCounter = () => {
  const [counter, setCounter] = useState(0)
  let i
  useEffect(() => {
    i = setInterval(() => {
      setCounter(c => c + 1)
    }, 500)
  }, () => {
    clearInterval(i)
  })
  return [counter, setCounter]
}

const App = ({ stdout }) => {
  const [counter] = useCounter()
  return (
    <Box height={stdout.rows} flexShrink={0} flexGrow={1} alignItems='center' justifyContent='center'>
      <Gradient name='rainbow'>
        <BigText text={`${counter} unicorns`} />
      </Gradient>
    </Box>
  )
}

export default withStdout(App)
```

And here is my `watcher.js`:
```js
import React from 'react'
import { watch } from 'chokidar'
import { render } from 'ink'
import clearRequire from 'clear-require'

let ready = false
let app

class AppWrapper extends React.Component {
  render () {
    const App = require('./App').default
    return <App />
  }
}

watch(__dirname, { ignored: /(^|[/\\])\../ })
  .on('all', (event, path) => {
    if (ready && (event === 'unlink' || event === 'add' || event === 'change')) {
      clearRequire.all()
      app.rerender()
    }
  })
  .on('ready', () => {
    ready = true
    app = render(<AppWrapper />, process.stdout)
  })
```

It works fine to render `App`, but with above watcher code, I get this error on change:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in App (created by Context.Consumer)
    in Unknown (created by AppWrapper)
    in AppWrapper
```
and the render crashes

**Comment by konsumer** [comment]
Here is same stuff in complete github project:
https://github.com/konsumer/ink-demo

**Comment by vadimdemedes** [comment]
Good idea! I'm currently focusing on resolving layout issues, but would certainly love to look into it later. Keep us posted with your progress!

**Comment by vadimdemedes** [comment]
Based on my investigation, HMR is only possible when you bundle with webpack or alike - something which majority of Ink-based CLIs don't.

I'm going to leave this issue open for visibility, but not planning to spend time on it myself for now.

**Comment by tom-sherman** [comment]
You could potentially try https://github.com/sidorares/hot-module-replacement so that a bundler isn't needed.

Also, if someone is gonna attempt this they should integrate with https://www.npmjs.com/package/react-refresh

**Comment by TheoBr** [comment]
Was looking into using Ink on a new project and was surprised this wasn't a thing yet. A simple ink vite template w/ hmr would be killer for greenfield and a useful resource for existing projects trying to improve their DX.

The starting point as things stand now with `npx create-ink-app` feels...weird? Like the release story is there but the local dev story is not

**Comment by tonyjara** [comment]
I know this is an old issue, but it came up first in google and I'm still learning ink. So, for whatever's worth, I'm just using bun with --watch to develop with ink and it works great.

**Comment by da1z** [comment]
hey. can anyone please give guidance how to achieve HMR? with vite or not does not matter. would really appreciate it

**Comment by da1z** [comment]
To whoever stumps upon this question I was able to make HMR work with vite-node (same package that powers vitest)

1. npm i -D vite vite-node @vitejs/plugin-react
2. create vite config
```
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";

export default defineConfig({
  plugins: [react()],
});

```
3. vite-node -w ./src/cli.tsx

that is it. you can run it and change code, everything reloads without react state loss. please let me know if it can be done better

**Comment by bgub** [comment]
@da1z I haven't been able to get this to work. Any tips?

**Comment by da1z** [comment]
> [@da1z](https://github.com/da1z) I haven't been able to get this to work. Any tips?

what is your issue? on my side my instructions worked from post above, i did not do anything special

check this https://github.com/google-gemini/gemini-cli/pull/4307/files

**Comment by jerriclynsjohn** [comment]
If anyone is looking around for a repo for this, here's one https://github.com/jerriclynsjohn/ink-cli-vite . Thank you @da1z and @bgub check this out, it works perfectly well!

**Comment by sindresorhus** [comment]
Closing as this is outside the scope of Ink itself. HMR is a tooling/bundler concern, not a UI framework concern. As shown in this thread, it's already achievable today using other tools.
